### PR TITLE
Albrja/feature/mic-6997/single jobmon UI

### PR DESCRIPTION
**Pull request by albrja**
### Albrja/feature/mic-6997/single jobmon UI
- *Category*: Refactor
- *JIRA issue*: https://jira.ihme.washington.edu/browse/MIC-6997

### Changes and notes
-refactor to use facade refactor pattern to have single place to interact with Jobmon UI

### Code reviewer
R Review: Albrja/feature/mic-6997/single jobmon UI
Confirmed PR target: base albrja/feature/mic-7114/handle-task-logs ← head albrja/feature/single-jobmon-ui (PR #338). 8 files, +188/-52, two commits (403f3d1 facade refactor + 07b03cd merge from base).

Summary
Introduces a new façade module [jobmon_config/client.py](vscode-webview://006i0prfhf5et33j75a80ran1o9buk0diqansps1hslm1ov3te5s/src/vivarium_cluster_tools/psimulate/jobmon_config/client.py) that centralizes all Jobmon SDK imports and calls. Direct jobmon.* imports and method calls in [workflow.py](vscode-webview://006i0prfhf5et33j75a80ran1o9buk0diqansps1hslm1ov3te5s/src/vivarium_cluster_tools/psimulate/jobmon_config/workflow.py), [runner.py](vscode-webview://006i0prfhf5et33j75a80ran1o9buk0diqansps1hslm1ov3te5s/src/vivarium_cluster_tools/psimulate/runner.py), [builder.py](vscode-webview://006i0prfhf5et33j75a80ran1o9buk0diqansps1hslm1ov3te5s/src/vivarium_cluster_tools/psimulate/workflow_config/builder.py), [interface.py](vscode-webview://006i0prfhf5et33j75a80ran1o9buk0diqansps1hslm1ov3te5s/src/vivarium_cluster_tools/psimulate/workflow_config/interface.py), and [task_builders.py](vscode-webview://006i0prfhf5et33j75a80ran1o9buk0diqansps1hslm1ov3te5s/src/vivarium_cluster_tools/psimulate/workflow_config/task_builders.py) are rerouted through client.* helpers (make_tool, make_task_template, make_workflow, create_task/create_tasks, add_upstream, add_tasks, bind_workflow, run_workflow, count_done). The two test files only retarget their mocker.patch strings at the new façade's Tool import site. No behavioral changes (verified — see Functionality below).

Design
create_task/create_tasks/add_tasks/add_upstream are one-line pass-throughs accepting **kwargs: Any ([client.py:79-97](vscode-webview://006i0prfhf5et33j75a80ran1o9buk0diqansps1hslm1ov3te5s/src/vivarium_cluster_tools/psimulate/jobmon_config/client.py#L79-L97), [:120-122](vscode-webview://006i0prfhf5et33j75a80ran1o9buk0diqansps1hslm1ov3te5s/src/vivarium_cluster_tools/psimulate/jobmon_config/client.py#L120-L122)) (design, maintainability). They re-export the same call with no validation, normalization, or default injection. Cost: mypy/IDE lose the precise signature at every caller (e.g. [task_builders.py:284](vscode-webview://006i0prfhf5et33j75a80ran1o9buk0diqansps1hslm1ov3te5s/src/vivarium_cluster_tools/psimulate/workflow_config/task_builders.py#L284) and [workflow.py:151](vscode-webview://006i0prfhf5et33j75a80ran1o9buk0diqansps1hslm1ov3te5s/src/vivarium_cluster_tools/psimulate/jobmon_config/workflow.py#L151) now get Any). Two options: (a) drop these wrappers and let callers call methods on the re-exported typed objects directly — the patch-target benefit is preserved by re-exporting Tool/Workflow; or (b) commit to a real domain API with explicit params (e.g. create_simulation_tasks(template, *, task_ids, metadata_dir, results_dir, max_attempts, command)).

Mixed parameter styles ([client.py:55-82](vscode-webview://006i0prfhf5et33j75a80ran1o9buk0diqansps1hslm1ov3te5s/src/vivarium_cluster_tools/psimulate/jobmon_config/client.py#L55-L82) vs [:79-97](vscode-webview://006i0prfhf5et33j75a80ran1o9buk0diqansps1hslm1ov3te5s/src/vivarium_cluster_tools/psimulate/jobmon_config/client.py#L79-L97)) (design, maintainability). make_task_template and make_workflow use explicit keyword-only params; create_task/create_tasks use **kwargs: Any. Asymmetry isn't documented as intentional. A Jobmon SDK change in create_task's signature would silently propagate, while a change in get_task_template would be caught.

bind_workflow returns "" as a "no URL" sentinel ([client.py:119-128](vscode-webview://006i0prfhf5et33j75a80ran1o9buk0diqansps1hslm1ov3te5s/src/vivarium_cluster_tools/psimulate/jobmon_config/client.py#L119-L128)) (design). Makes the return type a lie ("" is not a URL) and forces truthy-checks downstream. Optional[str] returning None is the standard Python idiom; the only caller [runner.py:79-83](vscode-webview://006i0prfhf5et33j75a80ran1o9buk0diqansps1hslm1ov3te5s/src/vivarium_cluster_tools/psimulate/runner.py#L79-L83) already does a truthy check that works either way.

run_workflow's RuntimeError guard is preserved-but-undocumented defensive code ([client.py:142-144](vscode-webview://006i0prfhf5et33j75a80ran1o9buk0diqansps1hslm1ov3te5s/src/vivarium_cluster_tools/psimulate/jobmon_config/client.py#L142-L144)) (design, documentation). Jobmon's typed contract for Workflow.run() says it returns str. This guard predates the PR (it was at the same spot in the old runner.py), so it's preserved behavior, but with the move it deserves either deletion or a comment linking to the version/incident that motivated it. Standalone defensive code with no rationale attracts cargo-cult propagation.

count_done is project domain logic in a Jobmon-SDK façade ([client.py:147-149](vscode-webview://006i0prfhf5et33j75a80ran1o9buk0diqansps1hslm1ov3te5s/src/vivarium_cluster_tools/psimulate/jobmon_config/client.py#L147-L149)) (design). The policy decision ("D" = done) is project-level; the function lives in a module nominally about Jobmon SDK encapsulation. Either move it back to its sole caller [runner.py:471](vscode-webview://006i0prfhf5et33j75a80ran1o9buk0diqansps1hslm1ov3te5s/src/vivarium_cluster_tools/psimulate/runner.py#L471), or rename/generalize to count_tasks_with_status(workflow, status) and surface "D" as a constant in the façade so [runner.py:148 / :477](vscode-webview://006i0prfhf5et33j75a80ran1o9buk0diqansps1hslm1ov3te5s/src/vivarium_cluster_tools/psimulate/runner.py#L148) (if wf_status != "D") can use the same symbol. Currently "D" is duplicated across three sites.

bind_workflow does two things ([client.py:119-128](vscode-webview://006i0prfhf5et33j75a80ran1o9buk0diqansps1hslm1ov3te5s/src/vivarium_cluster_tools/psimulate/jobmon_config/client.py#L119-L128)) (maintainability). Binds the workflow (state-changing) AND computes a URL string. At the call site [runner.py:79](vscode-webview://006i0prfhf5et33j75a80ran1o9buk0diqansps1hslm1ov3te5s/src/vivarium_cluster_tools/psimulate/runner.py#L79) it reads as if it's just binding. Suggest splitting: bind_workflow(workflow) -> None + get_monitoring_url(workflow) -> str | None.

Maintainability
Inconsistent verb conventions: make_* vs create_* ([client.py:50,56,94](vscode-webview://006i0prfhf5et33j75a80ran1o9buk0diqansps1hslm1ov3te5s/src/vivarium_cluster_tools/psimulate/jobmon_config/client.py#L50)) (maintainability). make_tool/make_task_template/make_workflow vs create_task/create_tasks. No semantic difference — both are factory wrappers. Future contributors have to guess which prefix to look for. Pick one.

count_done name is ambiguous at call site ([runner.py:471](vscode-webview://006i0prfhf5et33j75a80ran1o9buk0diqansps1hslm1ov3te5s/src/vivarium_cluster_tools/psimulate/runner.py#L471)) (maintainability). client.count_done(workflow) doesn't read as "count tasks with final_status==D". Rename to count_completed_tasks (or extract a module-level _JOBMON_STATUS_DONE = "D" and count_tasks_with_status(workflow, status)).

TOOL_NAME / CLUSTER_NAME are module-level but not in __all__ ([client.py:46-47](vscode-webview://006i0prfhf5et33j75a80ran1o9buk0diqansps1hslm1ov3te5s/src/vivarium_cluster_tools/psimulate/jobmon_config/client.py#L46-L47)) (maintainability, design). Tests or other modules that want to assert these constants have to reach in. Either add them to __all__ (if they're part of the contract) or prefix with _ (if internal). Documenting why "slurm" is hardcoded would help a future port.

Stale-ish comment ([runner.py:85-86](vscode-webview://006i0prfhf5et33j75a80ran1o9buk0diqansps1hslm1ov3te5s/src/vivarium_cluster_tools/psimulate/runner.py#L85-L86)) (maintainability). "Match the workflow timeout to the remaining time on the SLURM runner node so jobmon doesn't outlive (or underuse) the allocation" — the decision still belongs here, but the implementation moved into client.run_workflow. Either OK (the comment documents the call-site rationale) or move part of it into the façade docstring.

DRY
Repeated "build kwargs + conditionally add one optional" idiom ([client.py:64-82](vscode-webview://006i0prfhf5et33j75a80ran1o9buk0diqansps1hslm1ov3te5s/src/vivarium_cluster_tools/psimulate/jobmon_config/client.py#L64-L82), [:99-117](vscode-webview://006i0prfhf5et33j75a80ran1o9buk0diqansps1hslm1ov3te5s/src/vivarium_cluster_tools/psimulate/jobmon_config/client.py#L99-L117), [:137-150](vscode-webview://006i0prfhf5et33j75a80ran1o9buk0diqansps1hslm1ov3te5s/src/vivarium_cluster_tools/psimulate/jobmon_config/client.py#L137-L150)) (DRY). Three sites use the same shape. With only one optional kwarg per function, current code is arguably clearer than any helper. Watch, don't fix — extract if a fourth function or a second optional kwarg lands.

Patch-target string hand-duplicated across two test files ([test_workflow.py:32](vscode-webview://006i0prfhf5et33j75a80ran1o9buk0diqansps1hslm1ov3te5s/tests/psimulate/jobmon_config/test_workflow.py#L32), [test_builder.py:70](vscode-webview://006i0prfhf5et33j75a80ran1o9buk0diqansps1hslm1ov3te5s/tests/psimulate/workflow_config/test_builder.py#L70)) (DRY). Both hardcode "vivarium_cluster_tools.psimulate.jobmon_config.client.Tool". The whole point of the façade is single source of truth for patch targets, but the path is now hand-duplicated. Lift to a shared conftest.py (as either a constant or a shared mock_tool_cls fixture).

Constants consolidation is complete — verified "slurm" and "vivarium_cluster_tools" no longer appear as literals in src/ outside client.py. Pre-PR they were at multiple sites. ✓

Tests
149 new lines, 0 new tests ([client.py](vscode-webview://006i0prfhf5et33j75a80ran1o9buk0diqansps1hslm1ov3te5s/src/vivarium_cluster_tools/psimulate/jobmon_config/client.py)). No tests/psimulate/jobmon_config/test_client.py.

run_workflow None-status guard and seconds_until_timeout branch untested ([client.py:141-144](vscode-webview://006i0prfhf5et33j75a80ran1o9buk0diqansps1hslm1ov3te5s/src/vivarium_cluster_tools/psimulate/jobmon_config/client.py#L141-L144)) (tests). The conditional kwarg forwarding and the None raise have no direct test. Suggested tests in the sub-agent report — small, ~3 cases.

bind_workflow URL formatting and empty-config branch untested ([client.py:119-128](vscode-webview://006i0prfhf5et33j75a80ran1o9buk0diqansps1hslm1ov3te5s/src/vivarium_cluster_tools/psimulate/jobmon_config/client.py#L119-L128)) (tests). A typo in the URL template (missing # or wrong path) would silently produce broken links in slack notifications. ~2 tests recommended.

make_task_template / make_workflow "absent" branches of conditional kwargs untested ([client.py:74-75](vscode-webview://006i0prfhf5et33j75a80ran1o9buk0diqansps1hslm1ov3te5s/src/vivarium_cluster_tools/psimulate/jobmon_config/client.py#L74-L75), [:109-110](vscode-webview://006i0prfhf5et33j75a80ran1o9buk0diqansps1hslm1ov3te5s/src/vivarium_cluster_tools/psimulate/jobmon_config/client.py#L109-L110)) (tests). Existing tests cover only the "present" path (max_concurrently_running=10, default_compute_resources={...}). The "absent" path matters: [task_builders.py:_create_single_command_task](vscode-webview://006i0prfhf5et33j75a80ran1o9buk0diqansps1hslm1ov3te5s/src/vivarium_cluster_tools/psimulate/workflow_config/task_builders.py#L260) intentionally omits default_compute_resources, and a regression that always-inserts None would silently change SDK call shape.

count_done filter untested ([client.py:147-149](vscode-webview://006i0prfhf5et33j75a80ran1o9buk0diqansps1hslm1ov3te5s/src/vivarium_cluster_tools/psimulate/jobmon_config/client.py#L147-L149)) (tests). String-literal filter on "D" — exactly the kind of code that breaks silently when an SDK enum changes. Edge cases (empty workflow.tasks, mixed statuses) uncovered.

Existing test patch-target retargeting works but now exercises façade + caller together ([test_workflow.py:32](vscode-webview://006i0prfhf5et33j75a80ran1o9buk0diqansps1hslm1ov3te5s/tests/psimulate/jobmon_config/test_workflow.py#L32), [test_builder.py:70](vscode-webview://006i0prfhf5et33j75a80ran1o9buk0diqansps1hslm1ov3te5s/tests/psimulate/workflow_config/test_builder.py#L70)) (tests). Tests still pass because make_tool/make_task_template/make_workflow access Tool through client's module namespace, and tool_instance.get_task_template / .create_workflow mock chains survive the indirection. Acceptable, but no isolation between caller and façade — a façade-only bug needs façade-specific tests (items 15–18).

Documentation
Missing API reference RST for jobmon_config/ ([docs/source/api_reference/psimulate/](vscode-webview://006i0prfhf5et33j75a80ran1o9buk0diqansps1hslm1ov3te5s/docs/source/api_reference/psimulate/)) (documentation). Every other psimulate submodule has RST stubs (including the newly added workflow_config/*.rst), but no jobmon_config/ directory. The consolidation invariant ("All Jobmon imports and SDK calls live here") is invisible in rendered docs. Add jobmon_config/index.rst, client.rst, workflow.rst mirroring the workflow_config/ pattern.

Missing CHANGELOG.rst entry ([CHANGELOG.rst](vscode-webview://006i0prfhf5et33j75a80ran1o9buk0diqansps1hslm1ov3te5s/CHANGELOG.rst)) (documentation). The 3.2.0 - TBD block lacks an entry for this refactor. Patch-target relocation alone is maintainer-relevant. Suggested: - Consolidate Jobmon SDK calls into psimulate.jobmon_config.client (internal façade; patch targets in tests should now use this module).

bind_workflow docstring under-specifies side effects and config dependency ([client.py:125-134](vscode-webview://006i0prfhf5et33j75a80ran1o9buk0diqansps1hslm1ov3te5s/src/vivarium_cluster_tools/psimulate/jobmon_config/client.py#L125-L134)) (documentation). Should note (a) mutation via workflow.bind(), (b) URL is composed from JobmonConfig().get("http", "gui_url").

run_workflow docstring omits Raises and the seconds_until_timeout=None meaning ([client.py:137-150](vscode-webview://006i0prfhf5et33j75a80ran1o9buk0diqansps1hslm1ov3te5s/src/vivarium_cluster_tools/psimulate/jobmon_config/client.py#L137-L150)) (documentation). The RuntimeError contract is invisible from the docstring; seconds_until_timeout=None deferring to the SDK default isn't obvious.

Module docstring should call out TOOL_NAME/CLUSTER_NAME constants as canonical ([client.py:7-22](vscode-webview://006i0prfhf5et33j75a80ran1o9buk0diqansps1hslm1ov3te5s/src/vivarium_cluster_tools/psimulate/jobmon_config/client.py#L7-L22)) (documentation). One line so future contributors don't reintroduce string literals.

TODO removal in builder.py is correct ([builder.py:57 pre-PR](vscode-webview://006i0prfhf5et33j75a80ran1o9buk0diqansps1hslm1ov3te5s/src/vivarium_cluster_tools/psimulate/workflow_config/builder.py)) — the # TODO: MIC-6997 - encapsulate Jobmon UI in one place is appropriately removed; no other MIC-6997 refs in src/. ✓

Functionality
Behavior preserved across all five façade migration sites — verified by direct comparison:
bind_workflow: identical workflow.bind() + JobmonConfig().get(...) + URL formatting; empty-string fallback preserved.
run_workflow: identical conditional seconds_until_timeout kwarg forwarding + None-raise; just moved.
make_workflow: identical kwargs to tool.create_workflow (default_cluster_name="slurm", default_max_attempts, conditional max_concurrently_running).
make_task_template: identical kwargs to tool.get_task_template, conditional default_compute_resources.
count_done: identical filter expression.
JobmonConfigError was a dead import ([old runner.py:22](vscode-webview://006i0prfhf5et33j75a80ran1o9buk0diqansps1hslm1ov3te5s/src/vivarium_cluster_tools/psimulate/runner.py)) — imported but never caught anywhere in runner.py. The PR removes it; no behavioral regression. ✓

workflow.tasks.values() is the only workflow.tasks accessor in src/ — confirmed via grep. The façade fully encapsulates this attribute access. ✓

No silent type coercion or data transformation introduced — all helpers forward arguments verbatim.

Minor Nits
__all__ includes TaskTemplate ([client.py:31](vscode-webview://006i0prfhf5et33j75a80ran1o9buk0diqansps1hslm1ov3te5s/src/vivarium_cluster_tools/psimulate/jobmon_config/client.py#L31)) but no caller imports it. Either remove from __all__ or add a comment about it being part of the public surface for future helpers.

runner.py:76 docstring references "D" for DONE — fine, but ties the runner docstring to a Jobmon implementation detail that the façade is supposed to hide. Could just say "Jobmon status string (see Jobmon docs)".

Two RST ===== underlines in client.py:2-4 are width-mismatched to the title "Jobmon Client" — purely cosmetic but Sphinx may warn.

Overall
This is a clean and conservative refactor with no behavioral regressions and clear architectural intent. The façade succeeds at consolidating Jobmon imports and giving tests a single mock target.

Three categories of follow-up worth prioritizing:

High value, easy: add tests/psimulate/jobmon_config/test_client.py covering bind_workflow, run_workflow, count_done, and the conditional-kwarg branches of make_task_template / make_workflow (items 15–18, ~50 lines).
Medium value, easy: documentation gaps — CHANGELOG entry, Sphinx RST for jobmon_config/, Raises/side-effect docstrings (items 20–24).
Medium value, design call: revisit the four pure-pass-through wrappers (create_task/create_tasks/add_tasks/add_upstream) — either drop them and rely on the re-exported types for the mock target benefit, or commit to a real domain API. The current middle ground gives up type safety without earning abstraction (items 1, 2).
The "D" magic literal duplication (item 5) and bind_workflow's ""-vs-None choice (item 3) are worth fixing while the façade is still young — both compound as call sites grow.

### CODE REVIEW

**Review by patricktnast — APPROVED** [review]

File: src/vivarium_cluster_tools/psimulate/jobmon_config/workflow.py [hunk 1]
```diff
@@ -13,21 +13,17 @@
...
 from pathlib import Path
-from typing import TYPE_CHECKING
 
-from jobmon.client.api import Tool
 from loguru import logger
 
 from vivarium_cluster_tools.psimulate import TASK_RUNNER_MODULE, wrap_for_subprocess
 from vivarium_cluster_tools.psimulate.cluster.interface import NativeSpecification
+from vivarium_cluster_tools.psimulate.jobmon_config import client
+from vivarium_cluster_tools.psimulate.jobmon_config.client import Task, Tool, Workflow
```
Comment (rmudambi) [inline src/vivarium_cluster_tools/psimulate/jobmon_config/workflow.py]: Import these directly from Jobmon - can be in an if TYPE_CHECKING block?

File: src/vivarium_cluster_tools/psimulate/workflow_config/builder.py [hunk 2]
```diff
@@ -24,8 +23,7 @@
 from vivarium_cluster_tools.psimulate.workflow_config.utilities import is_resume
 
 if TYPE_CHECKING:
-    from jobmon.client.task import Task
-    from jobmon.client.workflow import Workflow
+    from vivarium_cluster_tools.psimulate.jobmon_config.client import Task, Workflow
```
Comment (rmudambi) [inline src/vivarium_cluster_tools/psimulate/workflow_config/builder.py]: Fix import here too.

File: src/vivarium_cluster_tools/psimulate/workflow_config/interface.py [hunk 3]
```diff
@@ -41,8 +41,7 @@
 )
 
 if TYPE_CHECKING:
-    from jobmon.client.api import Tool
-    from jobmon.client.task import Task
+    from vivarium_cluster_tools.psimulate.jobmon_config.client import Task, Tool
```
Comment (rmudambi) [inline src/vivarium_cluster_tools/psimulate/workflow_config/interface.py]: Same here

File: src/vivarium_cluster_tools/psimulate/runner.py [hunk 4]
```diff
@@ -51,15 +50,15 @@
 from vivarium_cluster_tools.vipin.perf_report import report_performance
 
 if TYPE_CHECKING:
-    from jobmon.client.workflow import Workflow
+    from vivarium_cluster_tools.psimulate.jobmon_config.client import Workflow
```
Comment (rmudambi) [inline src/vivarium_cluster_tools/psimulate/runner.py]: And here

File: tests/psimulate/jobmon_config/test_workflow.py [hunk 5]
```diff
@@ -28,8 +28,8 @@
 
 @pytest.fixture()
 def mock_tool_cls(mocker: MockerFixture) -> MagicMock:
-    """Patch the Jobmon ``Tool`` class at its import site."""
-    return mocker.patch("vivarium_cluster_tools.psimulate.jobmon_config.workflow.Tool")
+    """Patch the Jobmon ``Tool`` class at the façade's import site."""
+    return mocker.patch("vivarium_cluster_tools.psimulate.jobmon_config.client.Tool")
```
Comment (rmudambi) [inline tests/psimulate/jobmon_config/test_workflow.py:24]: Should need to update all these mocks too

**Review by rmudambi — APPROVED** [review]

File: tests/psimulate/jobmon_config/test_workflow.py (hunk 5 above)
Comment (albrja) [reply tests/psimulate/jobmon_config/test_workflow.py:24]: These are patches target the namespace so these were not distrubed.
